lib/decision/tls_identity.go

```diff
@@ -0,0 +1,273 @@
+		AWSRoleARNs:             id.AwsRoleArns,
+		AzureIdentities:         id.AzureIdentities,
+		GCPServiceAccounts:      id.GcpServiceAccounts,
+		ActiveRequests:          id.ActiveRequests,
+		DisallowReissue:         id.DisallowReissue,
+		Renewable:               id.Renewable,
+		Generation:              id.Generation,
+		BotName:                 id.BotName,
+		BotInstanceID:           id.BotInstanceId,
+		AllowedResourceIDs:      resourceIDsToTypes(id.AllowedResourceIds),
+		PrivateKeyPolicy:        keys.PrivateKeyPolicy(id.PrivateKeyPolicy),
+		ConnectionDiagnosticID:  id.ConnectionDiagnosticId,
+		DeviceExtensions:        deviceExtensionsFromProto(id.DeviceExtensions),
+		UserType:                types.UserType(id.UserType),
+	}
+}
+
+// TLSIdentityFromTLSCA transforms a [tlsca.Identity] into its equivalent
+// [decisionpb.TLSIdentity].
+func TLSIdentityFromTLSCA(id *tlsca.Identity) *decisionpb.TLSIdentity {
+	if id == nil {
+		return nil
+	}
+
+	return &decisionpb.TLSIdentity{
+		Username:                id.Username,
+		Impersonator:            id.Impersonator,
+		Groups:                  id.Groups,
+		SystemRoles:             id.SystemRoles,
+		Usage:                   id.Usage,
+		Principals:              id.Principals,
+		KubernetesGroups:        id.KubernetesGroups,
+		KubernetesUsers:         id.KubernetesUsers,
+		Expires:                 timestampFromGoTime(id.Expires),
+		RouteToCluster:          id.RouteToCluster,
+		KubernetesCluster:       id.KubernetesCluster,
+		Traits:                  traitFromWrappers(id.Traits),
+		RouteToApp:              routeToAppToProto(&id.RouteToApp),
+		TeleportCluster:         id.TeleportCluster,
+		RouteToDatabase:         routeToDatabaseToProto(&id.RouteToDatabase),
+		DatabaseNames:           id.DatabaseNames,
+		DatabaseUsers:           id.DatabaseUsers,
+		MfaVerified:             id.MFAVerified,
+		PreviousIdentityExpires: timestampFromGoTime(id.PreviousIdentityExpires),
+		LoginIp:                 id.LoginIP,
+		PinnedIp:                id.PinnedIP,
+		AwsRoleArns:             id.AWSRoleARNs,
+		AzureIdentities:         id.AzureIdentities,
+		GcpServiceAccounts:      id.GCPServiceAccounts,
+		ActiveRequests:          id.ActiveRequests,
+		DisallowReissue:         id.DisallowReissue,
+		Renewable:               id.Renewable,
+		Generation:              id.Generation,
+		BotName:                 id.BotName,
+		BotInstanceId:           id.BotInstanceID,
+		AllowedResourceIds:      resourceIDsFromTypes(id.AllowedResourceIDs),
+		PrivateKeyPolicy:        string(id.PrivateKeyPolicy),
+		ConnectionDiagnosticId:  id.ConnectionDiagnosticID,
+		DeviceExtensions:        deviceExtensionsToProto(&id.DeviceExtensions),
+		UserType:                string(id.UserType),
+	}
+}
+
+func timestampToGoTime(t *timestamppb.Timestamp) time.Time {
+	if t == nil {
+		return time.Time{}
+	}
+	return t.AsTime()
+}
+
+func timestampFromGoTime(t time.Time) *timestamppb.Timestamp {
+	if t.IsZero() {
```

Comment:
This is the one zero-value handling that seemed worth keeping, as I remember us having problems with the default timestamp before. Same for "timestampToGoTime".

Reply:
I had many issues because the default value for the go time and the timestamppb are different.

Timestamppb zero value is epoch, while go's zero is `0000-00-00 00:00:00 UTC`. I would recommend adding the epoch check in the `timestampToGoTime` to avoid this issue.

e.g.

```
func timestampToGoTime(t *timestamppb.Timestamp) time.Time {
	if t == nil {
		return time.Time{}
	}
	if goTime := t.AsTime(); goTime.Unix() != 0 {
		return goTime
	}
	return time.Time{}
}
```

This kind of problem is caused by some tooling (often terraform) creating the proto structs directly, they end up with their zero value, and we mis-convert in teleport.

Reply:
Done - 319caaa3c2b11a3a3f353b4789d58ae048286db5.